[Aio] Unary unary client call barebones implementation

pfreixes opened the pull request:
Implement the minimal stuff for making a unary call with the new
experimental gRPC Python implementation for `Asyncio`, called `Aio`.

What has been added:

- Minimal `iomgr` code for performing the required network and timer
calls.
- Minimal Cython code implementing the channel, call and the callback
context.
- Minimal Python code that mimics the synchronous implementation but
designed to be asynchronous.

Issues addressed:

- #19760 unary unary client call barebones implementation.
- #19793 When using a process executor a segmentation fault is received.

What is missing:

- Add the new test suite as part of the current CI.

Not solved yet:

- Tests have to be executed using the `GRPC_ENABLE_FORK_SUPPORT=0`
environment variable for skipping the fork handles installed by
the core library.

How is the `Aio` module used:

Following snippet shows how the `Aio` module can be used for making a unary call:

```python
async def coro():
    async with aio.insecure_channel(self.server_target) as channel:
        hi = channel.unary_unary(
            '/grpc.testing.TestService/UnaryCall',
            request_serializer=messages_pb2.SimpleRequest.SerializeToString,
            response_deserializer=messages_pb2.SimpleResponse.FromString
        )
        await hi(messages_pb2.SimpleRequest())
```

For running the tests, a new `test_aio` command has been added under the `setup.py` file. Tests can be executed using the `run_tests.py`. Following snippet shows the bash command used for compiling and running the tests:

```bash
$ GRPC_ENABLE_FORK_SUPPORT=0 python tools/run_tests/run_tests.py -l python-aio  --compiler python3.7
```

Co-authored-by: Manuel Miranda <manuel.miranda@skyscanner.net>
Co-authored-by: Mariano Anaya <mariano.anaya@skyscanner.net>
Co-authored-by: Zhanghui Mao <zhanghui.mao@skyscanner.net>

argaen commented on src/python/grpcio/grpc/experimental/aio/_channel.py:
```diff
@@ -0,0 +1,89 @@
+
+class UnaryUnaryMultiCallable(aio.UnaryUnaryMultiCallable):
+
+    # pylint: disable=too-many-arguments
```
how come do we need this here and not in the `Channel` constructor? (both have the same number of arguments)

pfreixes replied:
No idea, removing it since pylint is not complaining about this in my machine

pfreixes commented:
@gnossen @lidizheng ping

gnossen commented:
@pfreixes The title of this PR will be going into the release notes. Can we change it to something like "Add experimental support for asyncio unary client"?

gnossen commented:
Self-reminder that merging this PR is going to require an internal cherry-pick.

pfreixes commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2019 gRPC authors.
```
I will try it.

pfreixes commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi:
```diff
@@ -0,0 +1,149 @@
+
+        ops = <grpc_op *>gpr_malloc(sizeof(grpc_op) * _OP_ARRAY_LENGTH)
+
+        initial_metadata_operation = SendInitialMetadataOperation(_EMPTY_METADATA, GRPC_INITIAL_METADATA_USED_MASK)
```
I guess that by simply passing a metadata parameter that will be used later for building the operation. Do you anticipate a problem with that?

lidizheng replied:
Nope. We are going to implement it in the future anyway, to pass the tests.

pfreixes commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2019 gRPC authors.
```
I will try it

pfreixes commented on src/python/grpcio/grpc/experimental/aio/__init__.py:
```diff
@@ -0,0 +1,123 @@
+from grpc._cython.cygrpc import init_grpc_aio
+
+
+class Channel(six.with_metaclass(abc.ABCMeta)):
```
do you want to make it in this PR? could be done in a separated PR.

lidizheng replied:
Filed as https://github.com/grpc/grpc/issues/19975

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/sync_server.py:
```diff
@@ -0,0 +1,61 @@
+    can not coexist within the same process.
+    """
+
+    PORT = 3333
```
this would mean that this port would need to be sent back to the test since it is used for connecting to the server. Despite the changes required for coordinating the fixture and the test would not be rocket science, IMO they are still none negligible considering that this fixture will be removed once we have the asynchronous version of the server.

lidizheng replied:
Using a fixed port will introduce a flake in our tests, internally, they are ran in parallel. The port number can be conflicted. If you want to pick a random port, you can try this piece of code I wrote for testing. Or use the `portpicker` package.

```Python
@contextmanager
def get_free_loopback_tcp_port():
    if socket.has_ipv6:
        tcp_socket = socket.socket(socket.AF_INET6)
    else:
        tcp_socket = socket.socket(socket.AF_INET)
    tcp_socket.bind(('', 0))
    address_tuple = tcp_socket.getsockname()
    yield "localhost:%s" % (address_tuple[1])
    tcp_socket.close()
```

pfreixes replied:
Thanks, I will use this.

pfreixes commented:
Hi @lidizheng and @gnossen, I've identified the following issues with the CI checks regarding this PR. Any advice, guidance or event help from you will be welcomed:

1) `tools/distrib/yapf_code.sh` is failing [1] because it's using Python 2.7 as Python image for running the checks. it raises issues with Python 3 none compatibles keywords like `async`. Having the feeling that the way for addressing this would be moving the YAPF CI step to Python 3, rather than trying to hack the current implementation for having support for both environments.

2) `grpc_bazel_build` is also complaining [2] about the `ASYNCIO_SUPPORT` saying that this directive is not defined. Where can I define this besides of having it in the `setup.py` file?

3) Some tests are failing, for example, this one [3] and this other one [4]. I'm wondering if they are a blocker since they seem to not be required - am I getting in the right way the _REQUIRED_ word in the CI checks window?

Besides this, there is still pending:

1) Add the new Aio tests as part of the CI, we would need to execute the following command as a new entry of the CI checks: `GRPC_ENABLE_FORK_SUPPORT=0 python tools/run_tests/run_tests.py -l python-aio  --compiler python3.7` Who can take care of that?

2) Remember that there is still the restriction of using `GRPC_ENABLE_FORK_SUPPORT=0`, I've opened a discussion here #19961 for eventually get rid of that restriction. Who can participate on this discussion?

Thanks!


[1] https://source.cloud.google.com/results/invocations/78b81103-36b6-45c4-9dd6-ef558e04227a/targets/github%2Fgrpc%2Fsanity_linux_opt_native/tests;group=sanity_linux_opt_native;test=tools%2Fdistrib%2Fyapf_code.sh;row=2
[2] https://source.cloud.google.com/results/invocations/5882d5d6-7498-40af-8d90-ec38f3c436e8/targets/grpc%2Fcore%2Fpull_request%2Flinux%2Fgrpc_bazel_build/log
[3] https://source.cloud.google.com/results/invocations/1a2f39da-ebea-4a34-9b72-8ece247e2930/targets/github%2Fgrpc%2Faggregate_tests/tests 
[4]https://source.cloud.google.com/results/invocations/9ecdaf69-2f35-4d6f-a66d-6957ade5461e/targets/grpc%2Fcore%2Fpull_request%2Flinux%2Fgrpc_basictests_csharp/log

pfreixes commented:
@lidizheng everything that you commented has been addressed.

lidizheng commented:
@pfreixes  Can you take a look at the failing tests? Specifically, the Sanity Checks and Bazel Basic Tests for Python.

pfreixes commented:
Hi @lidizheng both test cases, if I'm not wrong, are part of the ones that I've commented in the previous comment:

- For the Bazel one is complaining about the `ASYNCIO_SUPPORT` directive [1], any hint about how to add and where to add this directive?

- For the Sanity one is complaining [2] about the Python 3 keywords not recognized by the YAPF Python 2 environment, what do you propose for addressing this issue?

[1] https://source.cloud.google.com/results/invocations/38568f5d-9ace-40ee-bfc5-ce9c6c74dff7/targets/grpc%2Fcore%2Fpull_request%2Flinux%2Fgrpc_bazel_build/log

[2] https://source.cloud.google.com/results/invocations/94fe8252-61e2-47a2-99b2-a88d55918fe9/targets/github%2Fgrpc%2Faggregate_tests/tests

lidizheng commented:
@pfreixes To supply `ASYNCIO_SUPPORT`, you can use the `--action_env` for Bazel. Check https://github.com/grpc/grpc/blob/master/tools/bazel.rc.

The sanity one, I can help.

pfreixes commented:
Thanks for the YAPF PR @lidizheng.

Regarding the inclusion of the `ASYNCIO_SUPPORT` directive, I'm a bit lost. Sure that there is something that I'm not getting in the right way.

I understood that the canonical way for building the Python package was by using the `setup.py` file provided here [1], indeed this is the file that is used for compiling and installing the package before running the tests [2].

The `ASYNCIO_SUPPORT` directive is figured out during the `setup.py` execution [3], and later exported [4] as a compile-time environment variable. If the `setup.py` file is not used, the variable won't be set and the Cython build will fail.

The Bazel build step that is failing - Bazel Basic build for C/C++ - is not using that `setup.py` file for compiling the package?

[1] https://github.com/grpc/grpc/blob/master/setup.py
[2] https://github.com/grpc/grpc/blob/master/tools/run_tests/helper_scripts/build_python.sh#L162
[3] https://github.com/Skyscanner/grpc/blob/unary_unary_client_aio/setup.py#L272
[4] https://github.com/Skyscanner/grpc/blob/unary_unary_client_aio/setup.py#L306

lidizheng commented:
Bazel won't use `setup.py` to build C-Extensions. I would say it requires advanced usage of Bazel to create a `config` or [`toolchain`](https://github.com/bazelbuild/bazel/issues/7375) for `asyncio`. @gnossen Do you have suggestions on how to fix the Bazel failure in this PR? I'm not super familiar with the new `toolchain` pattern.

gnossen commented:
I believe you'll just need to modify [this file](https://github.com/grpc/grpc/blob/master/tools/bazel.rc) a bit. I *think* you'll just need to add the lines

```
build --action_env=ASYNCIO_SUPPORT=""
```

and

```
build:python3 --action_env=ASYNCIO_SUPPORT=True
```

This will probably have to change with the next Bazel version bump, but it should be okay for now.

lidizheng commented:
@pfreixes As suggested by @gnossen in the previous PR, the include selection can be done within Cython layer. A short cut will be the [`version`](https://github.com/cython/cython/blob/master/Cython/Includes/cpython/version.pxd) module provided by Cython that allows you to do `IFDEF` on Cython source files. Can you try out this building method?

And the Bazel issue should goaway automatically, since the external dependency to compile time env is gone.

lidizheng commented:
After digging into the Cython compile time env, I found it seems not possible to use CPython macros in compile time. Unless we write an actual C file.

However, after experiments, the Cython understands `async`/`await` even in Python 2.7. That means we don't need to guard those files in compile time.

For example, an `async` function in Cython:
```Python
async def dance():
    pass
```

Will be compiled into:
```C++
PyObject *__pyx_gb_4grpc_7_cython_6cygrpc_84generator(__pyx_CoroutineObject *__pyx_generator, CYTHON_UNUSED PyThree
adState *__pyx_tstate, PyObject *__pyx_sent_value);
```

It uses generator, and returns coroutine. Both are defined in Python 2, so in the case of this PR, the `ASYNCIO_SUPPORT` is not needed.

---

That doesn't mean we can use `asyncio` in Python 2. We still need to detect whether `asyncio` is importable or not.

pfreixes commented:
Seems that you are right @lidizheng we do not need to be worried about compatibility within Cython, most likely because they implemented their own coroutine type so not having symbol issues if Python 2 is used with a Cython module compiled that requires a coroutine type.

Ive removed the compilation directive and I've just swallowed the import error when `asyncio` module cant be found.

In any case, I'm still considering that all of this stuff needs to be solved in a more holistic way and current status is kind of a workarround, I've added a TODO that points to that issue [1], I will update the issue later.

[1] https://github.com/grpc/grpc/issues/19728

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/iomgr.pyx.pxi:
```diff
@@ -12,7 +12,14 @@
+    # distributed without breaking none compatible Python versions. For now, if
+    # Asyncio package is not available we just skip it.
+    pass
+
```
Nit: all these `.pxi` files will be concatenated into one, so we can import the libraries once in `cygrpc.pyx`. I didn't mention it before because it doesn't hurt performance and increases readability. But this case we don't want to duplicate TODO comment and create burden (one change apply to multiple places) for future changes.

pfreixes replied:
Do you want to import the `asyncio` module just in one place? where?

lidizheng replied:
Yes, that's what I mean. In the top of `cygrpc.pyx`, there is multiple imports.
https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/cygrpc.pyx#L19

lidizheng commented on src/python/grpcio/grpc/_cython/cygrpc.pxd:
```diff
@@ -37,11 +37,10 @@ include "_cygrpc/grpc_gevent.pxd.pxi"
-    include "_cygrpc/aio/callbackcontext.pxd.pxi"
-    include "_cygrpc/aio/call.pxd.pxi"
-    include "_cygrpc/aio/channel.pxd.pxi"
+include "_cygrpc/aio/iomgr/socket.pxd.pxi"
```
nit: A short comment will help.

pfreixes replied:
what do you want to add?

lidizheng replied:
Something like: Following `.pxi` files are used for async API.
So they are logically separated from above `pxi` files.

pfreixes commented:
@lidizheng seems that I didn't remember to run the YAPF step in my local machine, syntax fixed with a new commit.

lidizheng commented:
@pfreixes To fix the Bazel failure, you can add the new `.pxi` files in https://github.com/grpc/grpc/blob/08ad670e55a31e507b4e90f3bea1a89041c44ca1/src/python/grpcio/grpc/_cython/BUILD.bazel#L5.

pfreixes commented:
Hi @lidizheng thanks for the comment, bazel file updated. Also, I've noticed that there was another bazel file for the experimental directory, I've added the missing entry for the new `aio` module. 

I did nothing for the new `tests_aio` directory regarding the Bazel configuration.

Also applied feedback related to how `asyncio` module is imported and added a short comment about the domain of the new included `pxi` files.

Let me know if this is enough.

pfreixes commented:
Besides the issue with the windows compilation, which should not be a big deal, I've found out that the tests sometimes got stuck when are executed using the command `GRPC_ENABLE_FORK_SUPPORT=0 python ./src/python/grpcio_tests/setup.py test_aio`. I've put the PR in WIP until this issue is solved.

pfreixes commented:
HI @lidizheng, please take a look at the last two commits [1] [2].

Regarding the issue with having the tests sometimes stuck when are executed all together by using the `setup.py` command - the issue can't be reproduced when the command used for running the tests is the regular `run_tests.py` -  I've found out that there was a segfault raised by the Glibc implementation of the `getaddrinfo` which is used by the server for resolving the address that has to be bound. I've circumvented the issue by using the `127.0.0.1` address taking into account that:

1) The Synchronous server fixture is a temporary thing that needs to be changed for an asynchronous version that will run within the same process, sharing the same loop. 
2) The scope of this PR is not having a 100% forking model working for all of the environments.
3) This issue might be an unsolvable issue if the root cause is a broken implementation of the `getaddrinfo` for MacOs after a `fork`.

[1] https://github.com/grpc/grpc/pull/19960/commits/8a80f8792a72b13ea60ded0a8385ec08fb91c863
[2] https://github.com/grpc/grpc/pull/19960/commits/991e35f79e1947693faceb0a3ec0fdb1562cdbfc

pfreixes commented:
Hi @lidizheng the issue is not specifically with windows, I faced the issue in my Mac when I was testing the conditional building for skipping the fork handlers.

This issue might be only specifically for Mac and any implementation using the `getaddrinfo` syscall might be implied.

pfreixes commented:
Seems that some tests failing for the Python Windows CI check step, but they are related to the gevent iomgr which is unrelated to the PR changes. The same for the other Ci checks that are failing.

Let me know if there is anything that I can do for unblocking this.

pfreixes commented on src/python/grpcio_tests/tests_aio/unit/test_base.py:
```diff
@@ -0,0 +1,53 @@
+        tcp_socket = socket.socket(socket.AF_INET)
+    tcp_socket.bind(('', 0))
+    address_tuple = tcp_socket.getsockname()
+    return tcp_socket, "127.0.0.1:%s" % (address_tuple[1])
```
good catch! lucky of us right now all of CI hosts used have dual-stack? In any case, we would need to fix this, I will work on this (not sure if Today or Tomorrow)

pfreixes commented:
@lidizheng & @gnossen squashed everything to only one commit, I'm planning on working on the nit advised by @gnossen but not sure when I will be able to do it, so it's up to you merge this or wait till I have the patch for supporting environments with only IPV6 support, which IMO might not be a blocker when most likely all CI environments supports dual-stack.

The good thing on having this merged is unblocking other stuff, like this future PR [1] or other issues like this one [2]

[1] https://github.com/Skyscanner/grpc/pull/9
[2] https://github.com/grpc/grpc/issues/20117

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi:
```diff
@@ -0,0 +1,149 @@
+        receive_status_on_client_operation.c()
+        ops[5] = <grpc_op> receive_status_on_client_operation.c_op
+
+        self._waiter_call = asyncio.get_running_loop().create_future()
```
`get_running_loop` is only available in Python 3.7.

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/resolver.pyx.pxi:
```diff
@@ -0,0 +1,61 @@
+    cdef void resolve(self, char* host, char* port):
+        assert not self._task_resolve
+
+        loop = asyncio.get_running_loop()
```
`get_running_loop` is only available in Python 3.7.

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/timer.pyx.pxi:
```diff
@@ -0,0 +1,45 @@
+        timer = _AsyncioTimer()
+        timer._grpc_timer = grpc_timer
+        timer._deadline = deadline
+        timer._timer_handler = asyncio.get_running_loop().call_later(deadline, timer._on_deadline)
```
`get_running_loop` is only available in Python 3.7.

lidizheng commented on src/python/grpcio/grpc/_cython/cygrpc.pxd:
```diff
@@ -36,3 +36,12 @@ include "_cygrpc/grpc_gevent.pxd.pxi"
 
```
Can you please help us to separate the C definition import from `grpc_gevent.pxd.pxi` to a separate file? Also, you were calling functions in `grpc_gevent.pyx.pxi` like `sockaddr_to_tuple`, can you also move them out to another file?

The new file can be named something like `iomgr.pxd.pxi`/`iomgr.pyx.pxi`, you decide.

This was a blocker because internally, we don't import `grpc_gevent` in some builds, and this PR will broke under those builds.

pfreixes replied:
addressed as was suggested.

pfreixes commented:
@lidizheng everything addressed.

@gnossen advice for improving what host is selected during the tests has been implemented here

[1] https://github.com/grpc/grpc/pull/19960/commits/26ce1721ae66fc00030653ade547b9189cdf5dff

pfreixes commented:
@lidizheng Updated BAZEL build file

lidizheng commented:
@pfreixes Seems the Bazel file still have syntax error:
```
ERROR: /var/local/git/grpc/src/python/grpcio/grpc/_cython/BUILD.bazel:44:9: unterminated string literal at eol
ERROR: /var/local/git/grpc/src/python/grpcio/grpc/_cython/BUILD.bazel:45:9: unterminated string literal at eol
ERROR: /var/local/git/grpc/src/python/grpcio/grpc/_cython/BUILD.bazel:45:1: Implicit string concatenation is forbidden, use the + operator
ERROR: /var/local/git/grpc/src/python/grpcio/grpc/_cython/BUILD.bazel:45:9: syntax error at '"_cygrpc/iomgr.pyd.pxi,"': expected ]
ERROR: /var/local/git/grpc/src/python/grpcio/grpc/_cython/BUILD.bazel:7:12: Traceback (most recent call last):
	File "/var/local/git/grpc/src/python/grpcio/grpc/_cython/BUILD.bazel", line 5
		pyx_library(name = "cygrpc", srcs = $error$, dep..."])
	File "/var/local/git/grpc/src/python/grpcio/grpc/_cython/BUILD.bazel", line 7, in pyx_library
		$error$
contains syntax error(s)
```

pfreixes commented:
fixed @lidizheng

lidizheng commented on src/python/grpcio/grpc/_cython/BUILD.bazel:
```diff
@@ -27,6 +41,8 @@ pyx_library(
         "_cygrpc/grpc_gevent.pyx.pxi",
         "_cygrpc/grpc_string.pyx.pxi",
+        "_cygrpc/iomgr.pxd.pxi",
+        "_cygrpc/iomgr.pyd.pxi",
```
s/pyd/pyx/
```
ERROR: missing input file '//src/python/grpcio/grpc/_cython:_cygrpc/iomgr.pyd.pxi'
ERROR: /var/local/git/grpc/src/python/grpcio/grpc/_cython/BUILD.bazel:5:1: //src/python/grpcio/grpc/_cython:cygrpc.pyx_cython_translation: missing input file '//src/python/grpcio/grpc/_cython:_cygrpc/iomgr.pyd.pxi'
ERROR: /var/local/git/grpc/src/python/grpcio/grpc/_cython/BUILD.bazel:5:1 1 input file(s) do not exist
```

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/iomgr.pxd.pxi:
```diff
@@ -0,0 +1,124 @@
+# Copyright 2017 gRPC authors.
```
`# Copyright 2019 The gRPC Authors`

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/iomgr.pyx.pxi:
```diff
@@ -0,0 +1,62 @@
+# Copyright 2018 gRPC authors.
```
`# Copyright 2019 The gRPC Authors`

pfreixes commented:
@lidizheng fixed last issues and everything rebased into one commit

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/resolver.pyx.pxi:
```diff
@@ -0,0 +1,61 @@
+        assert not self._task_resolve
+
+        loop = asyncio.get_event_loop()
+        self._task_resolve = asyncio.create_task(
```
Python 3.7 only feature.

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/socket.pyx.pxi:
```diff
@@ -0,0 +1,130 @@
+    cdef void connect(self, object host, object port, grpc_custom_connect_callback grpc_connect_cb):
+        assert not self._task_connect
+
+        self._task_connect = asyncio.create_task(
```
Python 3.7 only feature.

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/socket.pyx.pxi:
```diff
@@ -0,0 +1,130 @@
+    cdef void read(self, char * buffer_, size_t length, grpc_custom_read_callback grpc_read_cb):
+        assert not self._task_read
+
+        self._task_read = asyncio.create_task(
```
Python 3.7 only feature.

pfreixes commented:
Good catch @lidizheng, fixed and everything rebased in just one commit on top of the current master. Just in case, build a python 3.6 environment and tested everything under the python 3.6 version (something that I should have done from the very beginning)

```bash
(py36_native) $ python --version
Python 3.6.5
(py36_native) $ GRPC_ENABLE_FORK_SUPPORT=0 python ./src/python/grpcio_tests/setup.py test_aio
running test_aio
Running       _sanity._sanity_test.AioSanityTest.testTestsJsonUpToDate
Testing gRPC Python...
SUCCESS       _sanity._sanity_test.AioSanityTest.testTestsJsonUpToDate
Running       unit.channel_test.TestChannel.test_async_context
SUCCESS       unit.channel_test.TestChannel.test_async_context
Running       unit.channel_test.TestChannel.test_unary_unary
SUCCESS       unit.channel_test.TestChannel.test_unary_unary
Running       unit.init_test.TestInitGrpcAio.test_aio_supports_fork_and_grpc_aio_usable
SUCCESS       unit.init_test.TestInitGrpcAio.test_aio_supports_fork_and_grpc_aio_usable
Running       unit.init_test.TestInitGrpcAio.test_aio_supports_fork_and_grpc_blocking_usable
SUCCESS       unit.init_test.TestInitGrpcAio.test_aio_supports_fork_and_grpc_blocking_usable
Running       unit.init_test.TestInsecureChannel.test_insecure_channel
SUCCESS       unit.init_test.TestInsecureChannel.test_insecure_channel
6 tests finished:
	6 successful
	0 unsuccessful
	0 skipped
	0 expected failures
	0 unexpected successes
Interrupted Tests:
	[]

Errors/Failures:

Unexpected successes: []```

lidizheng commented on src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/iomgr.pyx.pxi:
```diff
@@ -0,0 +1,185 @@
+
+
+cdef void asyncio_socket_shutdown(grpc_custom_socket* grpc_socket) with gil:
+    raise NotImplemented()
```
Can you implement this function, it is required for internal build. Here is the version I wrote for tests to pass.
```Cython
cdef void asyncio_socket_shutdown(grpc_custom_socket* grpc_socket) with gil:
    socket = (<_AsyncioSocket>grpc_socket.impl)
    if socket.is_connected():
        socket._writer.close()
```

rmariano replied:
There is a slightly different implementation of this, already in https://github.com/Skyscanner/grpc/pull/8
It basically calls ``socket.close()`` as the ``socket.is_connected()`` verification is done internally.
Mentioning it in case it serves as reference, and also to watch out for conflicts.

lidizheng commented on src/python/grpcio_tests/tests_aio/unit/test_base.py:
```diff
@@ -0,0 +1,58 @@
+
+    @property
+    def server_target(self):
+        return self._target
```
Internally, there is some special optimization for test case running. If we setup a new process for each test case, the internal test runner will hang. So, I changed this class to initialize the process once per test class.
```Python
class AioTestBase(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        cls._socket, cls._target = _get_free_loopback_tcp_port()
        cls._server = sync_server.Server(cls._target)
        cls._server.start()
        cls._loop = asyncio.new_event_loop()
        asyncio.set_event_loop(cls._loop)
        aio.init_grpc_aio()

    @classmethod
    def tearDownClass(cls):
        cls._server.terminate()
        cls._socket.close()
        super().tearDownClass()

    @property
    def loop(self):
        return self._loop

    @property
    def server_target(self):
        return self._target
```

lidizheng commented:
After closely investigate the deadlock, I found the issue is quite complex. The trace log is [here](https://gist.github.com/lidizheng/820bcfa1e0d2b052e4e8280e418acad3).

1. Parent process A starts child process B;
1. Parent process A fires an RPC to process B, and succeed;
1. Parent process A sends SIGTERM to child process B;
1. Parent process A tries to fork child process C, two things possibly go wrong:
    1. Parent process A now is `grpc_init`ed, and C-Core has state;
        1. Since `GRPC_ENABLE_FORK_SUPPORT` is `false`, internal states won't be cleaned up.
        1. Leads to SEGFAULT
    1. Some metadata entries in parent process A is not fully cleaned up at this point;
        1. Reason unknown maybe the ref counter for metadata has a bug, or we are doing `grpc_shutdown` in a wrong way;
        1. Process A found there is a memory leak, aborted the `os.fork` operation;
1. `os.fork` failed, but the test continued;
1. Parent process A tries to fire another RPC to process C, which doesn't exist;
1. Parent process A deadlocks due to `asyncio` EventLoop selector not receiving any new events.

Link to [fork_support](https://github.com/grpc/grpc/blob/618a3f561d4a93f263cca23abad086ed8f4d5e86/doc/fork_support.md) document. If we must fork, I prefer the fork-before-use pattern with my modification. Or it might spending longer time to debug fork-while-using mechanism which I don't think is an effective path. The fork-while-using can be a future feature.

pfreixes commented:
Seems that we are still having issues for having forking support, which is a requirement because of how the server fixture is being implemented.

My proposal here will be, **remove everything that smells of having forking support** like fork handles and tests, and rewriting how the server fixture is being implemented by either having only one server instance for all of the tests or by forking always from a clean environment. TBH I would prefer that than having this kind of middle ground where seems that we would have support for forking but at the same time, we will be implementing this class attributes that circumvent forking issues with the internal gRPC google codebase.

Also, think that eventually the fixture will be replaced by a fully asynchronous server which will allow us to get rid of the spawned process.

WDYT @lidizheng?

lidizheng commented:
> My proposal here will be, **remove everything that smells of having forking support** like fork handles and tests, and rewriting how the server fixture is being implemented by either having only one server instance for all of the tests or by forking always from a clean environment.

I think your way is cleaner, I agree we should remove the forking support related code for now. About the rewrite, I suggest having a separate Python file for the server, and use `subprocess` to run it. Only the `test_base.py` and `sync_server.py` needs to be modified.

pfreixes commented:
@lidizheng forking support removed, the synchronous server is executed via `Popen` which runs the process using the fork and exec steps. The `GRPC_ENABLE_FORK_SUPPORT=0` is still needed for running the tests since the core is compiled by default having support for it.

lidizheng commented:
@pfreixes Seems YAPF is not happy.

pfreixes commented:
@lidizheng fixed, this is what happens when you run the linter after the YAPF :), let me know if you want to have everything squashed into one single commit.

lidizheng commented on tools/run_tests/run_tests.py:
```diff
@@ -882,6 +886,31 @@ def __str__(self):
+        if args.iomgr_platform not in ('native'):
+            raise Exception(
+                'Iomgr platform %s not supported.' % args.iomgr_platform)
+        return super()._get_pythons(args)
```
Strangely, this works in our CI (Kokoro). Eventually, we might want to migrate our test infra entirely to Python 3. But it would be better to use the Python 2 style (for now) to cope with other infra scripts.

pfreixes commented:
@lidizheng I agree that some comments are quite useless and they are not gonna help if you want to dig into the git history, but I thought that eventually everything will be squashed into one single commit with will come with a meaningful message - indeed every time that I'm squashing into one commit I'm reusing the first commit message.

Regarding the size of the commit, I agree as much smaller better. That's the reason that we tried to implement only the barebones of the unary call for the client, for making it as most simple we could. I hope that for next PRs related to the client the size of the PR will be more affordable since the foundations will be already there.

For the server-side, we might have the same issue since the first PR will come with the foundations for the server-side. Happy to discuss with you how can we break it down in smaller pieces but having yet the chance on making them testable.

So if you do not disagree, ping me for squashing everything into one commit.

lidizheng commented:
@pfreixes I'm facing another internal bug I couldn't fix with this new solution. It seems Google's CPython packaging is doing something weird, that the packaged binary doesn't work well with the test case process itself. I might dig deeper or work around it tomorrow, if you have any ideas about debugging this type of issue, please help.

Also, I think the server side of asyncio can have a prototype by the end of this week hopefully, if this PR can wait, I can try to check in both of them at the same time.

```
ImportError: ...net/grpc/python/_cython/cygrpc.so: undefined symbol: _ZN12DataPosition4PeekEv
```

pfreixes commented:
Weird, the error is a missing symbol from the cython binary file, how can
this be related with the tests?

I couldnt find the symbol in my local file, and neither a function in the
CPython core that might be related to it. Could this be related with some
google internals?

I guess that we always have the option of waiting for having the Server
ready, but it will limit the way we are implementing new features having
the requirement of implementing them for the server and the client. Commits
will be bigger for sure :)

Just realised that Ive forgotten to implement the close for the socket, I
will try to do it Today.

On Tue, Sep 10, 2019, 02:23 Lidi Zheng <notifications@github.com> wrote:

> @pfreixes <https://github.com/pfreixes> I'm facing another internal bug I
> couldn't fix with this new solution. It seems Google's CPython packaging is
> doing something weird, that the packaged binary doesn't work well with the
> test case process itself. I might dig deeper or work around it tomorrow, if
> you have any ideas about debugging this type of issue, please help.
>
> Also, I think the server side of asyncio can have a prototype by the end
> of this week hopefully, if this PR can wait, I can try to check in both of
> them at the same time.
>
> ImportError: ...net/grpc/python/_cython/cygrpc.so: undefined symbol: _ZN12DataPosition4PeekEv
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc/pull/19960?email_source=notifications&email_token=AAJLLCLT5IBW524L6SZBE73QI3SINA5CNFSM4IL3E3QKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6JNKAQ#issuecomment-529716482>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJLLCKNX3LN7BHV2JVGGE3QI3SINANCNFSM4IL3E3QA>
> .
>

pfreixes commented:
@lidizheng shutdown for `Aio` socket implemented, tell me if I can help you on anything for moving forward this PR.

If finally, the only way  - I hope no - for unblocking this PR is the server implementation, I'm eager to help you.

lidizheng commented:
@pfreixes I haven't got time today to actively work on this PR. I would say the fork-during-class-setup solution works best in both open source and Google internal. Do you think tradeoff some isolation between test methods is an option (rollback to previous solution)?

Since the fork happen before the traffic starts, it might not need those fork handlers.

pfreixes commented:
Oks @lidizheng but I'm still thinking that is a workaround since class methods are being called many times. But considering that this is will be removed once we have the asynchronous server we can do it.

In any case, seems that using the `setup.py` or the `run_tests.py` as entry points for running the tests, it does not work since the class methods are never called while running directly the test suite by using the raw `python ..../channel_test.py` works like a charm. 

Any idea why? you mentioned that you were able to run the tests.

lidizheng commented:
@pfreixes I only ran the tests using the new solution (subprocess). `run_tests.py` is another problem to be solved, you might need to take a look at the `loader.py`.

lidizheng commented:
@pfreixes Eventually, I found the root cause might lay deep in the building tool of Google. I adjusted how these aio tests are ran, and now it is passing. Can you revert the last commit `53c19332f19825ea34da220afcbccf8e63dcdbeb`, and squash the commits? If tests are happy, then we can merge it.

lidizheng commented:
@pfreixes It seems you touched one more commit... can you drop that commit?

pfreixes commented:
@lidizheng reverted, rebased and squashed into one single commit

lidizheng commented:
@pfreixes Thanks, running tests and the import script.

pfreixes commented:
any ETA for having it merged?
